deployer/src/project/docker.rs

```diff
@@ -141,26 +146,34 @@ pub struct ServiceDockerContext {
     docker: Docker,
     settings: Option<ContainerSettings>,
     runtime_manager: RuntimeManager,
+    events_tx: Arc<Mutex<Option<UnboundedSender<ProjectEvent>>>>,
 }
 
 impl ServiceDockerContext {
-    pub fn new(docker: Docker, runtime_manager: RuntimeManager) -> Self {
+    pub fn new(
+        docker: Docker,
+        runtime_manager: RuntimeManager,
+        events_tx: Arc<Mutex<Option<UnboundedSender<ProjectEvent>>>>,
```

Comment:
Is there anyway we can make the sender async so that it does not have to be behind a mutex?

Reply:
Not sure what you mean. The sender is already `async` (https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.UnboundedSender.html).

Reply:
Ahh :man_facepalming: 

But this does not have to be behind a `Mutex` right?

Reply:
Nvm, we will need the mutex to still reset the channel

Reply:
Then maybe we should have the `lock()` on the mutex usage do an immediate `clone()` to cause the guard to live as short as possible?

Reply:
I believe the guard already lives as short as possible. It should live as long as the operation involving the guarded object. In our case, we send an event after each service state transition, but also when creating the service (which doesn't correspond to a service state transition). Let me know if you thought about something else though.